torch/distributed/_shard/sharded_tensor/api.py

```diff
@@ -630,6 +630,8 @@ def cuda(
         return st_cuda
 
     def to(self, *args, **kwargs) -> ShardedTensor:
+        if not self._local_shards:
```

Comment:
I think this part could be slightly dangerous/tricky - for example the following:
st = ShardedTensor:
   rank 0 has shard_0
   rank 1 has no shards
 
we call st.to("cuda")

rank 0 will call a ShardedTensor.init_from_local_shards,
rank 1 will return self, 

this will cause rank 0 to block,

I think we may need to do a collective call the the beginning just to check if everyones device are the same as the passed in device. If not the same, then everyone needs to call init_from_local_shards,

cc @wanchaol

Reply:
I think this might be ok as `init_from_local_shards_and_global_metadata` is not calling collectives. But we need to align the behaviors indeed. `to` generally return a new Tensor, we should return a new ShardedTensor that have empty shards on this rank (with the same global sharded metadata, i.e. call `init_from_local_shards_and_global_metadata`), instead of returning self directly. This makes sure we have a different ShardedTensor object from both empty shards and non-empty shards. @colin2328

Reply:
@wanchaol , actually, I believe that if the current device and the to device are the same, it is a noop and returns the same tensor

>>> y = x.to("cuda")
>>> z = y.to("cuda")
>>> y.data_ptr() == z.data_ptr()
True

I believe the right thing to do is to return self in this case

Reply:
Oh that make sense, what I mean is that we should basically check `global_sharded_metadata`, and if every rank have empty shards, we can return self, but if some rank have empty shards while others not, we should return a new ShardedTensor with empty shards on those rank.

Reply:
ah, I see what you mean.